plugins/catalog-backend/src/ingestion/processors/AzureDevOpsDiscoveryProcessor.ts

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LocationSpec } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { ScmIntegrations } from '@backstage/integration';
+import { Logger } from 'winston';
+import * as results from './results';
+import { CatalogProcessor, CatalogProcessorEmit } from './types';
+import { codeSearch } from './azure';
+
+/**
+ * Extracts repositories out of an Azure DevOps org.
+ *
+ * The following will create locations for all projects which have a catalog-info.yaml
+ * on the default branch. The first is shorthand for the second.
+ *
+ *    target: "https://dev.azure.com/org/project"
+ *    or
+ *    target: https://dev.azure.com/org/project?path=/catalog-info.yaml
+ *
+ * You may also explicitly specify a single repo:
+ *
+ *    target: https://dev.azure.com/org/project/_git/repo
+ **/
+export class AzureDevOpsDiscoveryProcessor implements CatalogProcessor {
+  private readonly integrations: ScmIntegrations;
+  private readonly logger: Logger;
+
+  static fromConfig(config: Config, options: { logger: Logger }) {
+    const integrations = ScmIntegrations.fromConfig(config);
+
+    return new AzureDevOpsDiscoveryProcessor({
+      ...options,
+      integrations,
+    });
+  }
+
+  constructor(options: { integrations: ScmIntegrations; logger: Logger }) {
+    this.integrations = options.integrations;
+    this.logger = options.logger;
+  }
+
+  async readLocation(
+    location: LocationSpec,
+    _optional: boolean,
+    emit: CatalogProcessorEmit,
+  ): Promise<boolean> {
+    if (location.type !== 'azure-discovery') {
+      return false;
+    }
+
+    const azureConfig = this.integrations.azure.byUrl(location.target)?.config;
+    if (!azureConfig) {
+      throw new Error(
+        `There is no Azure integration that matches ${location.target}. Please add a configuration entry for it under integrations.azure`,
+      );
+    }
+
+    const { baseUrl, org, project, repo, catalogPath } = parseUrl(
+      location.target,
+    );
+    this.logger.info(
+      `Reading Azure DevOps repositories from ${location.target}`,
+    );
+
+    const files = await codeSearch(
+      azureConfig,
+      org,
+      project,
+      repo,
+      catalogPath,
+    );
+
+    this.logger.info(`Found ${files.length} files in Azure DevOps.`);
```

Comment:
Just noting that this will emit the log line on every refresh loop for every registered location. Looking at it from a log reader's point of view: this message should either be at debug level, or provide more information to give some context (the location being read, at least)